Doc/library/__main__.rst

```diff
@@ -1,25 +1,196 @@
 
 .. module:: __main__
-   :synopsis: The environment where the top-level script is run.
+   :synopsis: Command line interfaces, import-time behavior, and ``__name__ == '__main__'``
 
 --------------
 
-``'__main__'`` is the name of the scope in which top-level code executes.
-A module's __name__ is set equal to ``'__main__'`` when read from
-standard input, a script, or from an interactive prompt.
+In Python, ``__main__`` is not a single mechanism in the language, but in fact
+is part of two quite different constructs:
+
+1. The ``__name__ == '__main__'`` statement
+2. The ``__main__.py`` file in Python packages
+
+Each of these mechanisms are related to Python modules; how
+users interact with them and how they interact with each other. See
+section :ref:`tut-modules`.
+
+
+``__name__ == '__main__'``
+---------------------------
+
+``'__main__'`` is the name of the environment where top-level code is run.
+"Top-level code" means when a Python module is initialized from an interactive
+prompt, from standard input, from a file argument, from a :option:`-c`
+argument, or from a :option:`-m` argument, but **not** when it is initialized
+from an import statement.  In any of these situations, the module's
+``__name__`` is set equal to ``'__main__'``.
+
+The only other context in which Python code is run is when it is imported
+through an import statement. In that case, ``__name__`` is set equal to the
+module's name: usually the name of the file without the ``.py`` extension.
+
+As a result, a module can discover whether or not it is running in the
+top-level environment by checking its own ``__name__``, which allows a common
+idiom for conditionally executing code when the module is not initialized from
+an import statement::
+
+    if __name__ == '__main__':
+        # Execute when the module is not initialized from an import statement.
+        ...
+
+Idiomatic Usage
+^^^^^^^^^^^^^^^
+
+Putting as few statements as possible in the block below ``if __name___ ==
+'__main__'`` can improve code clarity. Most often, a function named *main*
+encapsulates the program's primary behavior, creating this pattern::
+
+    # echo.py
+
+    import shlex
+    import sys
+
+    def echo(phrase: str):
+       # you can imagine that this dummy wrapper around print might be
+       # different and truly worth re-using in a real program.
+       print(phrase)
+
+    def main():
+        "Echo the sys.argv to standard output"
+        echo(shlex.join(sys.argv))
+
+    if __name__ == '__main__':
+        main()
+
+This has the added benefit of the *echo* function itself being isolated and
+importable elsewhere::
+
+    # elsewhere.py
+
+    import sys
+
+    from echo import echo
+
+    def echo_platform():
+        echo(sys.platform)
+
+
+Packaging Considerations (``console_scripts``)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+For detailed documentation on Python packaging, see
+`setuptools. <https://setuptools.readthedocs.io/en/latest/>`__
```

Comment:
Please link to packaging.python.org instead.

Reply:
This change was implemented; see:

https://github.com/python/cpython/pull/26883/files#diff-ada229cf79d40b37dc14942330de96f7d48f2019cf0b1c864e8644765a4e3e7eR84-R85